ethcore/src/private_transactions/mod.rs

```diff
@@ -0,0 +1,819 @@
+						Some(account) => {
+							let accounts = self.accounts.read().clone().ok_or_else(|| PrivateTransactionError::AccountProviderIsMalformed)?;
+							let accounts = accounts.upgrade().ok_or_else(|| PrivateTransactionError::AccountProviderIsMalformed)?;
+							let private_state = self.execute_private_transaction(BlockId::Latest, &transaction)?;
+							let private_state_hash = keccak(&private_state);
+							trace!("Hashed effective private state for validator: {:?}", private_state_hash);
+							let signed_state = accounts.sign(account.address.clone(), account.password.clone(), private_state_hash)?;
+							let signed_private_transaction = SignedPrivateTransaction::new(desc.private_hash, signed_state, None);
+							trace!("Sending signature for private transaction: {:?}", signed_private_transaction);
+							self.broadcast_signed_private_transaction(signed_private_transaction.rlp_bytes().into_vec());
+						}
+						None => trace!("Cannot find validator account in config"),
+					}
+				},
+				Err(e) => trace!("Cannot retrieve descriptor for transaction with error {:?}", e),
+			}
+			self.transactions_for_verification.lock().remove_private_transaction(&transaction_hash, &fetch_nonce);
+		}
+		Ok(())
+	}
+
+	/// Add signed private transaction into the store
+	/// Creates corresponding public transaction if last required singature collected and sends it to the chain
+	pub fn import_signed_private_transaction(&self, rlp: &[u8]) -> Result<(), EthcoreError> {
+		let tx: SignedPrivateTransaction = UntrustedRlp::new(rlp).as_val()?;
+		trace!("Signature for private transaction received: {:?}", tx);
+		let private_hash = tx.private_transaction_hash();
+		if self.transactions_for_signing.lock().get(&private_hash).is_none() {
+			// Not our transaction, broadcast further to peers
+			self.broadcast_signed_private_transaction(rlp.into());
+			return Ok(());
+		}
+
+		let desc = self.transactions_for_signing.lock().get(&private_hash).expect("None was checked before; qed");
+		let last = self.last_required_signature(&desc, tx.signature())?;
+
+		if last {
+			let mut signatures = desc.received_signatures.clone();
+			signatures.push(tx.signature());
+			//Create public transaction
+			let public_tx = self.public_transaction(
+				desc.state.clone(),
+				&desc.original_transaction.clone(),
+				&signatures,
+				desc.original_transaction.nonce,
+				desc.original_transaction.gas_price
+			)?;
+			trace!("Last required signature received, public transaction created: {:?}", public_tx);
+			//Sign and add it to the queue
+			let accounts = self.accounts.read().clone().ok_or_else(|| PrivateTransactionError::AccountProviderIsMalformed)?;
+			let accounts = accounts.upgrade().ok_or_else(|| PrivateTransactionError::AccountProviderIsMalformed)?;
+			let chain_id = desc.original_transaction.chain_id();
+			let hash = public_tx.hash(chain_id);
+			let signer = self.config.lock().signer_account.clone();
```

Comment:
You'd better not to `clone()` the whole signer_account here — it should be enough to put a MutexGueard into a variable, and then clone only required fields out of it.

(This approach would work best with a new explicit lexical scope, out of which you would return a tuple `(address, password)`.)